### ISSUE TITLE
[16] Counterfactual Memorization in Neural Language Models

### ISSUE THREAD

**Issue by long8v** [long8v/PTIR#16]
<img width="754" alt="image" src="https://user-images.githubusercontent.com/46675408/160056323-1d6ad0f0-105f-45b3-8e63-a578d6613cb4.png">

[paper](https://arxiv.org/pdf/2112.12938.pdf)
Counterfactual memorization measures the expected change in a model’s prediction when a particular example is excluded from the training set. 
<img width="604" alt="image" src="https://user-images.githubusercontent.com/46675408/160054417-e8db1a2a-83f4-4329-ab64-2faf01459b66.png">
이를 구하기 위해서, 전체 트레이닝 데이터를 샘플링 한뒤 독립적으로 학습 시킨다. 각 모델들에 대해 특정 샘플 x가 있는 모델(이하 IN 모델)에서 next token prediction을 한 정확도와 샘플 x가 없는 모델(이하 OUT 모델)에서 같은 태스크를 했을 때의 정확도 차이를 측정한다.

More formally, we say a training example x is counterfactually memorized, when the model predicts x accurately if and only if the model was trained on x. 
1. We define counterfactual memorization in neural LMs which gives us a principled perspective to
distinguish “rare” (episodic) memorization from “common” (semantic) memorization in neural LMs
(Section 2).
2. We estimate counterfactual memorization on several standard text datasets, and confirm that rare
memorized examples exist in all of them. We study common patterns across memorized text in all
datasets and the memorization profiles of individual internet domains. (Section 3).
3. We extend the definition of counterfactual memorization to counterfactual influence, and study
the impact of memorized examples on the test-time prediction of the validation set examples and
generated examples (Section 4).

Generation-Time Memorization을 측정하기 위해서 generated 된 시퀀스가 training data에 있는지 확인하거나 training data를 구하기 어려우면 LM의 perplexity를 비교한다. 이러한 memorization과 우리가 제시하는 Counterfactual memorization의 다른 점은, 트레이닝 데이터에  near-duplicate 데이터가 많다면, 우리의 training set에서 subset을 제거하고 나서도 많이 남아있을 것이기 때문에 memorization으로 측정되지 않는다는 점이다.
In summary, generation-time memorization measures how likely a trained model would directly copy from the training examples, while counterfactual memorization aims to discover rare information that is memorized.

모델은 T5을 사용했고, C4, RealNews, Wiki를 사용했다. training data의 25%씩 subset으로 모델을 만들었다. 
<img width="863" alt="image" src="https://user-images.githubusercontent.com/46675408/160056001-c948a599-c42b-4149-be40-be8e6589454f.png">
실험 결과는 위와 같다. We found that examples with high memorization are generally unconventional text such as all-capital
letters, structured formats (i.e., tables or bullet list), and multilingual texts.
